peg/kotlin_post.h

```diff
@@ -0,0 +1,161 @@
+/*
+ *   Copyright (c) 2021 Jan Dolinár
+ *
+ *   This source code is released for free distribution under the terms of the
+ *   GNU General Public License version 2 or (at your option) any later version.
+ *
+ *   This module contains functions to generate tags for Kotlin.
+ */
+
+/*
+*   INCLUDE FILES
+*/
+#include "debug.h"
+#include "options.h"
+#include "parse.h"
+#include "routines.h"
+
+#include <signal.h>
+
+/*
+* FUNCTION DEFINITIONS
+*/
+static int getcFromKotlinFile(struct parserCtx *auxil)
+{
+    int c = getcFromInputFile();
+    if (auxil->parenthesis_level > 0 && (c == '\r' || c == '\n'))
+    {
+        return ' ';
+    }
+    return c;
+}
+
+
+static void pushKind (struct parserCtx *auxil, int kind)
+{
+    intArrayAdd (auxil->kind_stack, kind);
+}
+
+static void popScope(struct parserCtx *auxil)
+{
+    tagEntryInfo *e = getEntryInCorkQueue (auxil->scope_cork_index);
+    if (e)
+        auxil->scope_cork_index = e->extensionFields.scopeIndex;
+}
+
+static void popKind (struct parserCtx *auxil, bool popScopeToo)
+{
+    intArrayRemoveLast (auxil->kind_stack);
+
+    if (popScopeToo)
+    {
+        popScope(auxil);
+    }
+}
+
+static int peekKind (struct parserCtx *auxil)
+{
+    return intArrayLast (auxil->kind_stack);
+}
+
+static void reportError (struct parserCtx *auxil)
+{
+    auxil->found_syntax_error = true;
+    fprintf(stderr, "%s: syntax error in \"%s\"\n", getLanguageName (getInputLanguage ()), getInputFileName());
+}
+
+static void makeKotlinTag (struct parserCtx *auxil, const char *name, long offset, bool pushScope)
+{
+    int k = peekKind (auxil);
+    if (k == K_IGNORE) return;
+    tagEntryInfo e;
+    if (*name != '`')
+    {
+        initTagEntry(&e, name, k);
+    } else
+    {
+        size_t len = strlen(name) - 2;
+        char *stripped = (char *)PCC_MALLOC(auxil, len+1); //TODO: Should I free it? When?
+        memcpy(stripped, name + 1, len);
+        stripped[len] = '\0';
+        initTagEntry(&e, stripped, k);
+
+    }
+    e.lineNumber = getInputLineNumberForFileOffset (offset);
+    e.filePosition = getInputFilePositionForLine (e.lineNumber);
+    e.extensionFields.scopeIndex = auxil->scope_cork_index;
+    int scope_index = makeTagEntry (&e);
+    if (pushScope)
+    {
+        auxil->scope_cork_index = scope_index;
+    }
+}
+
+static void reportFailure(struct parserCtx *auxil, long offset)
+{
+    if(auxil->fail_offset < 0)
+    {
+        auxil->fail_offset = offset;
+    }
+}
+
+static void resetFailure(struct parserCtx *auxil, long offset)
+{
+    if(auxil->fail_offset >= 0)
+    {
+        unsigned long startLine = getInputLineNumberForFileOffset(auxil->fail_offset);
+        unsigned long endLine = getInputLineNumberForFileOffset(offset-1);
+        if (startLine == endLine)
+        {
+            fprintf(stderr, "Failed to parse '%s' at line %lu!\n", getInputFileName(), startLine);
```

Comment:
Just a question. Is this message for users or for debugging purposes?

In my idea, ctags should not report syntax errors in the input source files to USERS.
For debugging purposes, we can use macros defined in main/trace.h.
`TRACE_PRINT` is a suitable replacement for this fprintf.
To use trace, you have to run ./configure with `--enable-debugging`.
In addition, when you run ctags, you have to pass `--_trace=Kotlin`.

Reply:
It was meant for user. But if it is not common behavior in ctags, I'm okay with it.

I changed it to only track and print the parsing failures in debug mode.

Reply:
I must reconsider this topic in the future.